src/ILLink.RoslynAnalyzer/DynamicallyAccessedMembersAnalyzer.cs

```diff
@@ -53,33 +56,65 @@ public override void Initialize (AnalysisContext context)
 					}
 				}
 			});
-			// TODO: fix reporting for generic type substitutions. This should happen not only for method invocations,
-			// but for any reference to an instantiated method or type.
-			context.RegisterOperationAction (context => {
-				var invocationOperation = (IInvocationOperation) context.Operation;
-				ProcessInvocationOperation (context, invocationOperation);
-			}, OperationKind.Invocation);
+			context.RegisterSyntaxNodeAction (context => {
+				ProcessGenericParameters (context);
+			}, SyntaxKind.GenericName);
 		}
 
-		static void ProcessInvocationOperation (OperationAnalysisContext context, IInvocationOperation invocationOperation)
+		static void ProcessGenericParameters (SyntaxNodeAnalysisContext context)
 		{
-			if (context.ContainingSymbol.HasAttribute (RequiresUnreferencedCodeAnalyzer.RequiresUnreferencedCodeAttribute))
+			if (context.ContainingSymbol is not null && context.ContainingSymbol.HasAttribute (RequiresUnreferencedCodeAnalyzer.RequiresUnreferencedCodeAttribute)) {
 				return;
+			}
 
-			ProcessTypeArguments (context, invocationOperation);
-		}
-
-		static void ProcessTypeArguments (OperationAnalysisContext context, IInvocationOperation invocationOperation)
-		{
-			var targetMethod = invocationOperation.TargetMethod;
-			if (targetMethod.HasAttribute (RequiresUnreferencedCodeAnalyzer.RequiresUnreferencedCodeAttribute))
-				return;
+			ImmutableArray<ITypeParameterSymbol> typeParams = default;
+			ImmutableArray<ITypeSymbol> typeArgs = default;
+			if (context.SemanticModel.GetTypeInfo (context.Node).ConvertedType is INamedTypeSymbol type) {
```

Comment:
Why is the `ConvertedType` piece necessary?

If there is for example an implicit conversion from `GenericType<Foo>` to `OtherType<Bar>`, I think we want to check `GenericType<Foo>` here (but I might be wrong about what `ConvertedType` is - @agocke does that sound right to you?). We should have a test where the `ConvertedType` is not null to make sure we do the right thing.

Reply:
That was a copy and paste error, I don't need the implicit conversion, can do with just `Type` instead.